### ISSUE TITLE
MarkWrite trial labelling

### ISSUE THREAD

**Issue by Mark-Torrance** [isolver/OpenHandWrite#230]
This isn't really a bug - more an unwanted side effect of existing functionality. See initial discussion of automatic segmentation here: #95 

### Context and issue
For data collected using GetWrite, MarkWrite identifies trials just on the basis of start and end times, taken from the EXP_CV_1 table in the hdf5 file. Labelling is then provided by MarkWrite, giving Trial 1 to Trial N in temporal sequence.

Trials for which the are no pen samples do not (for obvious reasons) receive a label. That's correct (and necessary).

This has two consequences:
- For experiments where trial order is randomised, different trials (different stimulus presentations) get the same trial label.
- Trials "go missing".  Imagine an experiment has three trials, perhaps labelled by the user and mytrial1, mytrial2 and mytrial3, presented in that order. If a participant provides a pen response for all three trials, then they will be labelled Trial 1, Trial 2, and Trial 3. If they only provide a pen response in mytrial1 and mytrial3, these will get labelled Trial 1 and Trial 2.

This doesn't affect the usability of the resulting report. Summarising by trial and them merging into EXP_CV_1 with a left join, matching using DV_TRIAL_ID, works fine. But for some (most?) experimental tasks this makes markup confusing.

### Suggested solution
Allow users to name a trial label column. This defaults to DV_TRIAL_ID but can be any user-provided trial label that appears in the EXP_CV_1 table. This can be set in the ioHub HDF5 Trial Segmentation options.

Sensible users will then provide their own trial label which (rather than rely on labelling by sequence in a specific session).

**Comment by Mark-Torrance** [comment]
Thinking more about this: It's actually a more serious issue. The current automatic top-level segment labelling doesn't, of course, just affect what is displayed in the MarkWrite GUI. It also provide that top-level label in the output. If this label is used in subsequent analysis to identify trial, then skipping one trial will mean that all subsequent trial numbers are wrong. This can be avoided simply by not relying on the trial label (by merging with the EXP_CV_1 data table and using user-provided trial labels). But it would be helpful to use these user-provided trial labels at the top-level segment labels in the first place.

**Comment by isolver** [comment]
I feel like this is in part because there is confusion about currently what a MarkWrite trial _label_ is, which is not equal to the trial's id. It is also clear that this confusion is unneeded and should/could be improved.

> Allow users to name a trial label column. This defaults to DV_TRIAL_ID but can be any user-provided trial label that appears in > the EXP_CV_1 table. This can be set in the ioHub HDF5 Trial Segmentation options.

My concern about leaving this up to the user is that it does not ensure each trial label is unique. I wonder if we should just switch it back to how it originally worked, using DV_TRIAL_ID to generate the N in a "Trial N" label. So if there are 4 trials and trial 3 has no data, the trial labels should be:

```
Trial 1
Trial 2
Trail 4
```

vs the current:

```
Trial 1
Trial 2
Trail 3
```

By using DV_TRIAL_ID, which is set by the experiment at the end of each trial to the current trial index, we should be ensuring that trial labels are unique,  while also remove the confusion around trial labels looking like trial ids but not actually being the trial id. ;).

What do you think?

>  This can be avoided simply by not relying on the trial label (by merging with the EXP_CV_1 data table and using user-provided trial labels). 

Is this not already being done by the reports? Merging with EXP_CV should be occurring automatically already, so using not using the trial _label_ created by Markwrite, and instead using the actual trial id or other column from EXP_CV during analysis, should not be that big of an issue. For example the sample report includes the EXP_CV data for each sample, including the ROW_INDEX and DV_TRIAL_ID columns.

**Comment by Mark-Torrance** [comment]
> By using DV_TRIAL_ID, which is set by the experiment at the end of each trial to the current trial index, we should be ensuring that trial labels are unique, while also remove the confusion around trial labels looking like trial ids but not actually being the trial id. ;).
> 
> What do you think?

Yes. This is all that's needed. The deluxe fix might be to allow users to add a tag to the trial label from another column in EXP_CV. This is, actually, pretty essential if, for example, experiments are set up with more than one pen-capture event per trial. But also might be useful for markup in some contexts. I'm just thinking of `Trial1_usertag1` etc.

> Is this not already being done by the reports? Merging with EXP_CV should be occurring automatically already, so using not using the trial label created by Markwrite, and instead using the actual trial id or other column from EXP_CV during analysis, should not be that big of an issue. For example the sample report includes the EXP_CV data for each sample, including the ROW_INDEX and DV_TRIAL_ID columns.

A problem arises when some trials don't include pen data. These trials then (necessarily) disappear in the existing segment and sample reports. That's fine. Users just need to be aware of this. It might be worth creating a custom report that just outputs EXP_CV. But actually this is also easily handled in R (I'll write something for the wiki). At some point we need to stop trying to guess what researchers will want and pass responsibility on to them.
